lua: search modules next to the main script

Totktonada opened the pull request:
It eliminates a need to call `package.setsearchroot()` in the main script to find modules in the application's directory.

Such layout is typical for cartridge based applications. It it also used, when different versions of modules are in use in different applications and installing the modules into the system is not an option.

Let's assume that there is an application in the directory /path/to/myapp with modules inside:

```
+ /path/to/myapp/
  +- .rocks/share/tarantool/
     +- foo.lua
  +- init.lua
  +- myapp/
     +- bar.lua
```

Now the following command automatically adds the path to the main script directory to search paths for modules:

```sh
$ tarantool /path/to/app/init.lua
```

So `require('foo')` and `require('myapp.bar')` works without any additional work.


Follows up #7774
Fixes #8182

Buristan reviewed (commented):
Hi, Alexander!
Thanks for the patch-set!
I'll proceed with review per-patch.

**[PATCH 1/2] lua: add minifio.script() to get main script path**
LGTM.

Buristan commented on src/lua/loaders.lua:
```diff
@@ -250,6 +250,47 @@ table.insert(package.loaders, 5, gen_loader_func(search_rocks_lib, load_lib))
+        ROCKS_LIB_TEMPLATES)
+
+    -- Mix the script directory loaders into corresponding
+    -- searchroot based loaders. It allows to don't change
```
Minor: I suggest rephrasing it like the following:
`It allows to avoid changing ordinals of the loaders ...`
The original comment is hard for me to read:(.

Totktonada replied:
No problems, changed.

Buristan commented on test/app-luatest/module_search_test.lua:
```diff
@@ -0,0 +1,94 @@
+    local module_slug = module_relpath
+        :gsub('^%.rocks/share/tarantool/', 'rocks/'):gsub('/', '_'):sub(1, -5)
+
+    g['test_' .. module_slug] = function(g)
```
Side note: Why "slug"?:)

Totktonada replied:
TBH, I don't know from where it comes.

https://www.npmjs.com/package/slugify
https://pypi.org/project/python-slugify/
https://github.com/sindresorhus/slugify

Buristan replied:
Interesting, thanks!:)

Buristan commented on src/lua/loaders.lua:
```diff
@@ -250,6 +250,47 @@ table.insert(package.loaders, 5, gen_loader_func(search_rocks_lib, load_lib))
 
+-- Search for modules next to the main script.
+local script = minifio.script()
+if script ~= nil and script ~= '-' then
```
Side note: This part gave me some interesting ideas to explore:

```sh
$ cat $PWD/-
print("test")
$ luajit -e 'print("-e")' -- -
-e
test
$ lua -e 'print("-e")' -- -
-e
test
$ src/tarantool -e 'print("-e")' -- -
-e
^C^C^C^C^C^C^Z
[1]+  Stopped                 src/tarantool -e 'print("-e")' -- -
```

`--`: stops handling options in Lua/LuaJIT (see [Lua 5.1 Reference Manual](https://www.lua.org/manual/5.1/manual.html#6))
As far as it is not implemented via Tarantool we can ignore this case, for now.

Totktonada replied:
No problems.

```diff
diff --git a/src/lua/loaders.lua b/src/lua/loaders.lua
index 2d7cb69b0..794f74eb5 100644
--- a/src/lua/loaders.lua
+++ b/src/lua/loaders.lua
@@ -251,6 +251,10 @@ table.insert(package.loaders, 5, gen_loader_func(search_rocks_lib, load_lib))
 -- croot          8
 
 -- Search for modules next to the main script.
+--
+-- If the script is not provided (script == nil) or provided as
+-- stdin (script == '-'), there is no script directory, so nothing
+-- to do here.
 local script = minifio.script()
 if script ~= nil and script ~= '-' then
     -- It is important to obtain the directory at initialization,
```

Buristan reviewed (approved):
**[PATCH 2/2] lua: search modules next to the main script**
LGTM, just a few reflections below.

igormunkin reviewed (approved):
**[PATCH 1/2] lua: add minifio.script() to get main script path**

@Totktonada, thanks for the patch! LGTM, but I'm disappointed with the way CLI arguments are handled by Tarantool.

```
$ cat t.lua
print(arg[0])
$ ./src/tarantool t.lua 
t.lua
$ ./src/tarantool -e 'print(arg[0])'
./src/tarantool
$ luajit t.lua
t.lua
$ luajit -e 'print(arg[0])'
nil
```

I have no idea, why Tarantool handles arguments this way; LuaJIT approach allows to solve the original issue without `minifio.script` helper (hence this patch is not needed).

Anyway, no changes are required within this patch.

Totktonada commented:
In turn, I dislike some aspects of how LuaJIT forms `arg` (see #7936). Anyway, I think that it is good to have an internal helper that is independent from our position on how `arg` should be interpreted in different contexts. (Maybe it worth to cover `minifio.script()` with a test, hm. I'll add a task to my wishlist.)

igormunkin reviewed (approved):
**[PATCH 2/2] lua: search modules next to the main script**

@Totktonada, thanks for the patch! LGTM except some nits, I've left below.

Typo: s/It it/It is/.

Minor: there is not a single word regarding the rationale for such a complex change instead of simply patching `package.path`/`package.cpath` and `LUA_PATH`/`LUA_CPATH`. I suggest to left the comment both in the commit message and in `src/lua/loaders` nearby the code patching the default loaders.

Totktonada commented on test/app-luatest/module_search_test.lua:
```diff
@@ -0,0 +1,94 @@
+    -- See also override_test.lua.
+    {'override/socket.lua', 'socket'},
+    {'.rocks/share/tarantool/override/socket.lua', 'socket'},
+}) do
```
A kind of my pattern: wrap everything related to a test case into a table. Sometimes it has named fields, sometimes just two items. I would leave it as present.

Totktonada commented:
> Minor: there is not a single word regarding the rationale for such a complex change instead of simply patching `package.path`/`package.cpath` and `LUA_PATH`/`LUA_CPATH`. I suggest to left the comment both in the commit message and in `src/lua/loaders` nearby the code patching the default loaders.

In fact, I would rather implement all the loaders as package.{path,cpath} generators, but while we have them as loaders, the new logic is easier to add as loader wrappers. (See also a discussion in #3136).

I have added a comment into the code.

<details>
<summary>The change (collapsed).</summary>

```diff
diff --git a/src/lua/loaders.lua b/src/lua/loaders.lua
index 23d294cee..794f74eb5 100644
--- a/src/lua/loaders.lua
+++ b/src/lua/loaders.lua
@@ -277,6 +277,17 @@ if script ~= nil and script ~= '-' then
     -- searchroot based loaders. It allows to avoid changing
     -- ordinals of the loaders and also makes the override
     -- loader search here.
+    --
+    -- We can just add more paths to package.path/package.cpath,
+    -- but:
+    --
+    -- * Search for override modules is implemented as a loader.
+    -- * Search inside .rocks in implemented as a loaders.
+    --
+    -- And it is simpler to wrap this logic rather than repeat.
+    -- It is possible (and maybe even desirable) to reimplement
+    -- all the loaders logic as paths generation, but we should
+    -- do that for all the logic at once.
     package.loaders[2] = chain_loaders({
         package.loaders[2],
         gen_loader_func(search_app_lua, load_lua),
```
</details>

Added a remark into the commit message:

> Regarding the implementation. There are two possible ways to add more searching logic: add more entries into `package.path` and `package.cpath` or add/wrap a loader. I'm going the second way, because we already have a logic around .rocks and override modules implemented as loaders. It is easier to follow.
>
> However, I would reimplement all those loaders as paths generators for `package.path` and `package.cpath` in a future. See also #3136.
